src/analysis/souffle/taint.dl

```diff
@@ -70,21 +83,30 @@
 // that source path. Instead, we remove all of the tags on the source from the
 // midpoint access path we add in the middle of each edge. This rule populates a
 // request to remove a tag on the edge midpoint.
-//      - it is safe because these requests are only accepted if they would 
+//      - it is safe because these requests are only accepted if they would
 //      normally be by the removeTag(_, _) rule
-//      - it does not require the manifest writer to know anything about the 
+//      - it does not require the manifest writer to know anything about the
 //      tags in the analysis
-//      - data owners do not need to know about paths, and they will grant 
-//      permission for edge removal indirectly by granting permission to remove 
+//      - data owners do not need to know about paths, and they will grant
+//      permission for edge removal indirectly by granting permission to remove
 //      tags (abstract representations of their data).
 
-saysRemoveTag(principal, midpoint, tag) :-
-  isTag(tag), claimNotEdge(principal, src, tgt), edgeToMidpointAccessPath(src, tgt, midpoint).
+ownsAccessPath(principal, path) :- saysOwnsAccessPath(principal, principal, path).
+ownsAccessPath(principal, tgt) :-
+  resolvedEdge(principal, src, tgt), ownsAccessPath(principal, src).
+
+// saysRemoveTag(principal, midpoint, principal, tag) :-
+//   isTag(tag), claimNotEdge(principal, src, tgt), edgeToMidpointAccessPath(src, tgt, midpoint).
+
+removeTag(path, owner, tag) :-
+   saysRemoveTag(owner, path, owner, tag).
+
+hasTag(path, owner, tag) :- saysHasTag(owner, path, owner, tag).
 
-mayHaveTag(tgt, tag) :- hasTag(tgt, tag).
+mayHaveTag(tgt, owner, tag) :- hasTag(tgt, owner, tag).
 
-mayHaveTag(tgt, tag) :-
-    resolvedEdge(src, tgt), mayHaveTag(src, tag), !removeTag(tgt, tag).
+mayHaveTag(tgt, owner, tag) :-
+    resolvedEdge(owner, src, tgt), mayHaveTag(src, owner, tag), !removeTag(tgt, owner, tag).
```

Comment:
OK. I think this makes sense to me and I think this answer is an improvement on what we have.

Here's my explanation of what you're doing:
In this rule, the `owner` is the owner of `tag` (in all positions in which it appears). `resolvedEdge` is the set of edges the `owner` principal believes exist. By default, `resolvedEdge` believes all the edges produced by the app graph (in `edge`) are present. Edges can be removed if the `owner` permits this by delegating the right to remove edges to some other principal (e.g., `owner says PrincipalX canSay claimNotEdge(...); PrincipalX says claimNotEdge(...)`).

At the system design / conceptual / explainability level, this seems pretty reasonable.

(If we wanted to keep open the opportunity of defining and proving IFC theorems, this will be more plausible if we can iterate on this again and pack all the ways of removing tags into a single downgrades relation. I don't have a concrete proposal for that though).

Reply:
For context in case this changes, `resolvedEdge` is currently populated by:

```
resolvedEdge(principal, src, tgt) :-
   isPrincipal(principal),
   edge(src, tgt),
   !claimNotEdge(principal, src, tgt).
   ```

The above means "A `resolvedEdge` for `principal` corresponds to an `edge` from `src` to `tgt` where `principal` did not claim that the edge does not exist".

On this line, we are saying that a `tgt` `mayHaveTag` `(owner, tag)` if there exists an `edge` that `owner` did not claim does not exist and `tgt` does not remove the tag `(owner, tag)`.

This seems narrower than what we currently have. What if a Principal other than the tag owner claims that an edge does not exist? As currently written, the `owner` would only believe that the edge is not there if the owner themselves claimed it. In reality, there are probably multiple principals that the owner is willing to believe about their dataflow graph, and we would want to exclude the `edge`s claimed to not exist by those principals as well.

Reply:
> As currently written, the owner would only believe that the edge is not there if the owner themselves claimed it. In reality, there are probably multiple principals that the owner is willing to believe about their dataflow graph, and we would want to exclude the edges claimed to not exist by those principals as well.

 Well, the owner can write `"OtherPrincipal" canSay claimNotEdge(src, tgt)` and the owner will believe it when other principal removes the edge (and this also allows `"OtherPrincipal"` to delegate it further). This is consistent with how things tend to work with authorization logic.

Reply:
> What if a Principal other than the tag owner claims that an edge does not exist?

In that case, the tag will not be propagated along the edge unless the owner trusts the principal. Isn't that the intended behavior.  The delegation will happen along the same lines as `removeTag`:

```
// "EndUser" trusts "EFF" to manage edge removal on its behalf.
"EndUser" says "EFF" canSay claimNotEdge("EndUser", src, tgt) :- 
    isAccessPath(src), isAccessPath(tgt).
// "EFF" audits "P" and trusts the edge removal in particle "P".
"EFF" says "P" canSay claimNotEdge(o, src, tgt) :- 
   isPrincipal(o), isAccessPath(src), isAccessPath(tgt).
// "P" makes the clam about edge.
says_claimNotEdge("P", "owner", "P#0.foo", "P#0.bar").
```

All of this will result in `claimNotEdge("owner", "P#0.foo", "P#0.bar")`.

Reply:
I agree with what Gogul said here.

Reply:
It sounds like this is relying upon us having `says_canSay_someFact(speaker, delegatee, ...)` cause the `speaker` to "echo" the matching statement by `delegatee`. I'm still a little bit concerned about this echoing pattern causing confusion as to who actually said a particular fact and correlating the attributes from a claim to the speaker, but @aferr showed me some workarounds, so I think I'm satisfied for now.

As you all have pointed out, as long as we believe the echoing pattern is fine, so is this line of code, so I'm convinced.

Reply:
The "echo"-ing pattern that you refer to is actually an implementation detail that should not be visible to the end user as long as the authorization logic syntax makes sense.

However, I also feel that "says" or "canSay" does not capture the intent quite well I think.  Instead of
`"EndUser" says "EFF" canSay claimNotEdge("EndUser", src, tgt):- ...`, the following might be more readable:

```
"EndUser" trusts "EFF" ifItSays claimNotEdge("EndUser", src, tgt) :- 
    isAccessPath(src), isAccessPath(tgt).
```

(Or something like that.) When we use `"canSay"`, it leaves things incomplete. e.g., "EFF" can say `claimNotEdge("EndUser", src, tgt)`. So, what are the implications if "EFF" says it.  Does this resonate with your concern, @markww? 

In any case, this is somewhat orthogonal issue involving the front-end syntax and we should sort this out later.